File: code/1690. Stone Game VII.py
class Solution:
    def stoneGameVII(self, stones: List[int]) -> int:
        n = len(stones)
        Sum = [0]
        for s in stones:
            Sum.append(Sum[-1] + s)
        v = [[-1 for i in range(n)] for j in range(n)]
        def dp(l, r):
            if l == r:
                return 0
            if v[l][r] != -1:
                return v[l][r]
            
            ans = max(Sum[r + 1] - Sum[l + 1] - dp(l + 1, r), Sum[r] - Sum[l] - dp(l, r - 1))
            v[l][r] = ans
            return v[l][r]
        return dp(0, n - 1)

File: code/1690. Stone Game VII.cpp
class Solution {
public:
    int stoneGameVII(vector<int>& stones) {
        int n = stones.size();
        vector<int> Sum(n + 1, 0);
        for (int i = 0; i < n; i++) {
            Sum[i + 1] = Sum[i] + stones[i];
        }
        vector<vector<int>> memo(n, vector<int>(n, INT_MAX));
        function<int(int, int)> dp = [&](int l, int r) {
            if (l == r) return memo[l][r] = 0;
            if (memo[l][r] != INT_MAX) return memo[l][r];
            int diff_l = dp(l + 1, r);
            int diff_r = dp(l, r - 1);
            return memo[l][r] = max(Sum[r + 1] - Sum[l + 1] - diff_l, Sum[r] - Sum[l] - diff_r);
        };
        return dp(0, n - 1);
    }
};